### PR TITLE
[SYCL-MLIR] Attach fast math flags to codegened operations

### PR DESCRIPTION
**Pull request by victor-eds**
Attach fast math flags to:
- Binary arithmetic operations
- Unary arithmetic operations
- `math.fma` operations
- `sycl.math.*` operations

### CODE REVIEW

File: polygeist/tools/cgeist/Test/Verification/sycl/fastmath.cpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+// CHECK-FAST:        %[[VAL_171:.*]] = arith.subf %[[VAL_169]], %[[VAL_170]] fastmath<fast> : f32
+// CHECK-NOHONOR:     %[[VAL_171:.*]] = arith.subf %[[VAL_169]], %[[VAL_170]] fastmath<nnan,ninf> : f32
+// CHECK:             return %[[VAL_171]] : f32
+// CHECK:           }
+
+// CHECK:           func.func @_Z3fmaIfET_S0_S0_S0_(
+// CHECK-SAME:                                      %[[VAL_175:.*]]: f32 {llvm.noundef}, %[[VAL_176:.*]]: f32 {llvm.noundef}, %[[VAL_177:.*]]: f32 {llvm.noundef}) -> (f32 {llvm.noundef})
+// CHECK-PRECISE:     %[[VAL_178:.*]] = math.fma %[[VAL_175]], %[[VAL_176]], %[[VAL_177]] : f32
+// CHECK-FAST:        %[[ADD:.*]] = arith.mulf %[[VAL_175]], %[[VAL_176]] fastmath<fast> : f32
+// CHECK-FAST:        %[[VAL_178:.*]] = arith.addf %[[ADD]], %[[VAL_177]] fastmath<fast> : f32
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/sycl/fastmath.cpp:97]: Surprised `-ffast-math` does this. I check and `llvm` is doing the same, tho

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLMathOps.td [hunk 2]
```diff
@@ -40,7 +41,10 @@ def SYCLMathFunc : SYCLOpTrait<"SYCLMathFunc">;
 // TODO: Support marrarys.
 // TODO: Consider VectorUnrollOpInterface, ElementwiseMappable.
 class SYCL_MathOp<string mnemonic, list<Trait> traits = []> :
-    SYCL_Op<"math." # mnemonic, traits # [SYCLMathFunc, Pure]>;
+    SYCL_Op<"math." # mnemonic,
+            traits # [SYCLMathFunc,
+                      DeclareOpInterfaceMethods<ArithFastMathInterface>,
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLMathOps.td:46]: This is needed to get the `getFastMathAttrName()` static member function

File: mlir-sycl/lib/Conversion/SYCLToMath/SYCLToMath.cpp [hunk 3]
```diff
@@ -44,7 +44,7 @@ struct OneToOneMappingPattern : public OpConversionPattern<SYCLOpT> {
     // `op` has a native MLIR type, hence we can just replace it with its
     // counterpart in the `math` dialect.
     NamedAttribute fastmath = rewriter.getNamedAttr(
-        MathOpT::getFastMathAttrName(), adaptor.getFastmathAttr());
+        SYCLOpT::getFastMathAttrName(), adaptor.getFastmathAttr());
```
Comment (victor-eds) [inline mlir-sycl/lib/Conversion/SYCLToMath/SYCLToMath.cpp:47]: Equivalent, but conceptually better

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 4]
```diff
@@ -2793,7 +2803,7 @@ ValueCategory MLIRScanner::EmitBinDiv(const BinOpInfo &Info) {
             << "Not applying OpenCL/HIP precision options.\n";
       }
     });
-    return LHS.FDiv(Builder, Loc, RHS);
+    return LHS.FDiv(Builder, Loc, RHS, getFastMathFlags(Info.getFPFeatures()));
```
Comment (sommerlukas) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2806]: Nit: The expression `getFastMathFlags(Info.getFPFeatures())` appears quite often, would it make sense to introduce a helper for it?

**Review by sommerlukas — APPROVED** [review]

**Review by sommerlukas — APPROVED** [review]

**Pull request merged**